README.md
# StateX

O StateX tem como fundamento principal a padronização da gerência de estado de maneira 
simples e centralizada em uma store de gerenciamento. Outra vantagem desse tipo de abordagem é o ganho de performance, pois os builders trabalham de maneira a qual atende um escopo específico evitando uma mudança em toda a árvore de widgets.

### Nada melhor que um exemplo para entender o funcionamento

```dart 
//Repositório mockado para exemplificação.
class Repository {
  Future<bool> authUser(String email, String passwold) async {
    await Future.delayed(Duration(seconds: 3));
    return true;
  }
}

//Dessa forma eu é criado um novo view model onde será centralizado do gerenciamento de estado da home
class HomeViewModel extends StateXStore {
  //Dessa maneira é criado um novo atributo a ter o estado gerenciado pelo StateX
  late final xEmail = StateX.of(this)("");
  late final xPassword = StateX.of(this)("");

  final repo = Repository();

  set email(String value) => xEmail.value = value;
  String get email => xEmail.value;

  set password(String value) => xPassword.value = value;
  String get password => xPassword.value;

  bool get isValid => email.isNotEmpty && password.isNotEmpty;

  void auth() async {
    //Aqui é informado que a view estará em loading
    setLoading(true);
    if (await repo.authUser(email, password)) {
      //Aqui é informado que a view que o loading pode ser revogado
      setLoading(false);
    } else {
      //Em caso de erro assim deve ser setado a exceção
      setError(Exception("Email or password is invalid"));
    }
  }
}

void main() {
  //Aqui observer para todos os estados da aplicação útil para gerenciamento de erros ou logs dos estados da aplicação
  StateXGlobalObserver.listen((type, state) => print(type));
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'StateX',
      theme: ThemeData(primarySwatch: Colors.blue, accentColor: Colors.white),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  var viewModel = HomeViewModel();

  @override
  void initState() {
    //Aqui é adicionado um novo observer útil para gerenciamento de erros, contudo qualquer estado que seja mudado dentro da store passará por aqui
    viewModel.addObserver((type, state, _) => print(type));
    super.initState();
  }

  @override
  dispose() {
    viewModel.dispose();
    super.dispose();
  }

  Widget _buildButton(bool inLoading) {
    Widget buttonChild = inLoading
        ? SizedBox(
            width: 20,
            height: 20,
            child: CircularProgressIndicator(
              color: Colors.white,
            ))
        : Text("Enter");
    return Container(
      width: double.infinity,
      height: 48,
      child: ElevatedButton(
        onPressed: viewModel.isValid ? viewModel.auth : null,
        child: buttonChild,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("StateX"),
        ),
        body: Padding(
          padding: const EdgeInsets.all(21),
          child: Center(
            child: SingleChildScrollView(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text(
                    'Login do usuário',
                  ),
                  TextField(
                    onChanged: (value) {
                      viewModel.email = value;
                    },
                  ),
                  SizedBox(
                    height: 20,
                  ),
                  TextField(
                    onChanged: (value) {
                      viewModel.password = value;
                    },
                  ),
                  SizedBox(
                    height: 20,
                  ),
                  //Exemplo usando  scoped builder
                  StateXScopedBuilder(
                    //É informado qual a store que gerenciará as mudanças nesse escopo
                    store: viewModel,
                    //Parâmetro opcional. Especifica qual atributos irá escutar as mudanças de estado, caso não passe este atributo o scoped rebuildará a cada mudança da store
                    states: [viewModel.xEmail, viewModel.xPassword],
                    onLoading: () => _buildButton(true),
                    onState: () => _buildButton(false),
                  ),
                  SizedBox(
                    height: 20,
                  ),
                  //Exemplo usando state builder
                  StateXBuilder(
                    //É informado qual a store que gerenciará as mudanças nesse builder
                    store: viewModel,
                    //Parâmetro opcional. Especifica qual atributos irá escutar as mudanças de estado, caso não passe este atributo o scoped rebuildará a cada mudança da store
                    states: [viewModel.xEmail, viewModel.xPassword],
                    //Parâmetro informo que será rebuildado quando a store estiver loading
                    includeLoading: true,
                    builder: (context, snapshot) {
                      return _buildButton(
                          snapshot.type == StateXType.isLoading);
                    },
                  )
                ],
              ),
            ),
          ),
        ));
  }
}

```


CHANGELOG.md
## 1.0.2
 * Change all states after set load false.
## 1.0.1
 * Change all states after set load false.
## 1.0.0
 * Fisrt release.


test/src/stores/state_x_store_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:state_x/src/stores/state_x_store.dart';
import 'package:state_x/src/utils/state_x.dart';

class StoreTest extends StateXStore {
  late final state = StateX.of(this)(0);
}

void main() {
  test('State x store observer add and remove', () async {
    StateXType type = StateXType.isState;
    var store = StoreTest();
    var obs = (tp, state, ownerCode) => type = tp;
    store.addObserver(obs);
    store.setLoading(true);
    expect(type == StateXType.isLoading, true);

    type = StateXType.isState;
    store.removeObserver(obs);
    store.setLoading(true);
    expect(type == StateXType.isLoading, false);
  });

  test('State x store observer change state', () async {
    var store = StoreTest();
    int stateValue = 0;
    var obs = (tp, state, ownerCode) =>
        tp == StateXType.isState ? stateValue = state : "";
    store.addObserver(obs);
    store.state.value = 20;
    expect(stateValue, 20);
  });

  test('State x store observer set error', () async {
    var store = StoreTest();
    Exception error = Exception();

    var obs =
        (tp, state, ownerCode) => tp == StateXType.isError ? error = state : "";

    store.addObserver(obs);
    store.setError(Exception("123456"));

    expect(error.toString(), "Exception: 123456");
  });
}


test/src/builders/state_x_builder_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:state_x/state_x.dart';

class StoreTest extends StateXStore {
  late final state = StateX.of(this)(0);
}

var store = StoreTest();

class TestApp extends StatefulWidget {
  @override
  _TestAppState createState() => _TestAppState();
}

class _TestAppState extends State<TestApp> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StateXBuilder(
        store: store,
        builder: (context, snapshot) {
          if (snapshot.type == StateXType.isLoading) {
            return Text("StateLoading ${snapshot.data}");
          } else if (snapshot.type == StateXType.isError) {
            return Text("StateError ${snapshot.data}");
          } else if (snapshot.type == StateXType.isState) {
            return Text("IsState");
          }
          return Container();
        },
      ),
    );
  }
}

void main() {
  testWidgets('State x builder state flux', (tester) async {
    await tester.pumpWidget(MaterialApp(
      home: TestApp(),
    ));

    expect(find.text("IsState"), findsOneWidget);

    store.setLoading(true);
    await tester.pump();
    expect(find.text("StateLoading true"), findsOneWidget);

    store.setLoading(false);
    await tester.pump();
    expect(find.text("IsState"), findsOneWidget);

    store.setError(Exception("123456"));
    await tester.pump();
    expect(find.text("StateError Exception: 123456"), findsOneWidget);

    store.state.value = 20;
    await tester.pump();
    expect(find.text("IsState"), findsOneWidget);
  });
}


test/src/builders/state_x_scoped_builder_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:state_x/src/stores/state_x_store.dart';
import 'package:state_x/state_x.dart';

class StoreTest extends StateXStore {
  late final state = StateX.of(this)(0);
}

var store = StoreTest();

class TestApp extends StatefulWidget {
  @override
  _TestAppState createState() => _TestAppState();
}

class _TestAppState extends State<TestApp> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StateXScopedBuilder(
        store: store,
        onState: () => Text("IsState"),
        onError: (error) => Text("StateError $error"),
        onLoading: () => Text("StateLoading"),
      ),
    );
  }
}

void main() {
  testWidgets('State x scoped builder state flux', (tester) async {
    await tester.pumpWidget(MaterialApp(
      home: TestApp(),
    ));

    expect(find.text("IsState"), findsOneWidget);

    store.setLoading(true);
    await tester.pump();
    expect(find.text("StateLoading"), findsOneWidget);

    store.setLoading(false);
    await tester.pump();
    expect(find.text("IsState"), findsOneWidget);

    store.setError(Exception("123456"));
    await tester.pump();
    expect(find.text("StateError Exception: 123456"), findsOneWidget);

    store.state.value = 20;
    await tester.pump();
    expect(find.text("IsState"), findsOneWidget);
  });
}


lib/src/stores/state_x_store.dart
import '../utils/state_x_global_observer.dart';

enum StateXType { isLoading, isState, isError }

class StateXLoading {
  final bool inLoading;

  StateXLoading(this.inLoading);
}

typedef StateXObserver = Function(
    StateXType type, Object? state, int ownerCode);

typedef StateXNotifierObserver = Function(Object? state, int key,
    {String? observerKey});

class StateXStore {
  Exception? error;
  bool inLoading = false;

  final Map<String, StateXObserver> _observers = <String, StateXObserver>{};
  StateXNotifierObserver get stateXObserver => _stateXObserver;

  _stateXObserver(Object? state, int key, {String? observerKey}) {
    StateXType type = StateXType.isState;
    if (state is StateXLoading) {
      type = StateXType.isLoading;
      inLoading = state.inLoading;
    } else if (state is Exception) {
      type = StateXType.isError;
      error = state;
    } else {
      error = null;
      inLoading = false;
    }

    if (_observers.isNotEmpty) {
      if (observerKey != null && _observers[observerKey] != null) {
        _observers[observerKey]!(type, state, key);
      } else {
        for (var obs in _observers.entries) {
          obs.value(type, state, key);
        }
      }
    }
    StateXGlobalObserver.dispatch(type, state);
  }

  void setLoading(bool status, {String? key}) {
    _stateXObserver(StateXLoading(status), status.hashCode, observerKey: key);
  }

  void setError(Exception error, {String? key}) {
    _stateXObserver(error, error.hashCode, observerKey: key);
  }

  void addObserver(StateXObserver observer, {String? key}) {
    var observerKey = (key ?? observer.hashCode.toString());
    _observers.addAll({observerKey: observer});
  }

  void removeObserver(StateXObserver observer) {
    _observers.remove(observer.hashCode.toString());
  }

  void removeObserverKey(String key) {
    _observers.remove(key);
  }

  void dispose() {
    _observers.clear();
  }
}


lib/src/builders/state_x_builder.dart
import 'package:flutter/material.dart';
import '../stores/state_x_store.dart';

typedef CallBuilder = Widget Function(
    BuildContext context, FluxSnapshot snapshot);

class FluxSnapshot<T> {
  final StateXType type;
  final T data;

  FluxSnapshot(this.type, this.data);
}

class StateXBuilder extends StatefulWidget {
  final StateXStore store;
  final CallBuilder builder;
  final List<Object>? states;
  final bool includeLoading;
  const StateXBuilder({
    required this.store,
    required this.builder,
    this.states,
    this.includeLoading = true,
  });

  @override
  _StateXBuilderState createState() => _StateXBuilderState();
}

class _StateXBuilderState extends State<StateXBuilder> {
  late Widget Function(BuildContext cxt) _currentBuilder;
  //StateXType? _lastStateType;
  bool _inLoading = false;
  void _observer(StateXType type, Object? state, int ownerCode) {
    if (widget.states != null && type == StateXType.isState) {
      var contains =
          widget.states!.where((rx) => rx.hashCode == ownerCode).isNotEmpty;
      if (!contains) {
        return;
      }
    }

    var _changeState = true;
    var _builder;
    switch (type) {
      case StateXType.isLoading:
        _changeState = widget.includeLoading;
        var status = (state as StateXLoading).inLoading;
        _inLoading = status;
        if (status) {
          _builder = (ctx) =>
              widget.builder(context, FluxSnapshot<bool>(type, status));
        } else {
          _builder = (ctx) =>
              widget.builder(context, FluxSnapshot(StateXType.isState, null));
        }
        break;
      case StateXType.isError:
        _changeState = true;
        _builder = (ctx) => widget.builder(
            context, FluxSnapshot<Exception?>(type, state as Exception?));
        break;
      default:
        if (_inLoading) {
          _changeState = false;
        }
        _builder = (ctx) => widget.builder(context, FluxSnapshot(type, null));
    }

    if (_changeState) {
      setState(() {
        _currentBuilder = _builder;
      });
    }
  }

  @override
  void initState() {
    _currentBuilder =
        (ctx) => widget.builder(ctx, FluxSnapshot(StateXType.isState, null));

    widget.store.addObserver(_observer);
    super.initState();
  }

  @override
  void dispose() {
    widget.store.removeObserver(_observer);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return _currentBuilder(context);
  }
}


lib/src/builders/state_x_scoped_builder.dart
import 'package:flutter/material.dart';
import '../stores/state_x_store.dart';

typedef CallState = Widget Function();
typedef CallError = Widget Function(Exception error);
typedef CallLoading = Widget Function();

class StateXScopedBuilder extends StatefulWidget {
  final StateXStore store;
  final CallState onState;
  final CallError? onError;
  final CallLoading? onLoading;
  final List<Object>? states;

  StateXScopedBuilder({
    required this.store,
    required this.onState,
    this.states,
    this.onLoading,
    this.onError,
  });

  @override
  _State createState() => _State();
}

class _State extends State<StateXScopedBuilder> {
  Widget _currentResult = Container();
  bool _inLoading = false;

  void _observer(StateXType type, Object? state, int ownerCode) {
    if (widget.states != null && type == StateXType.isState) {
      var contains =
          widget.states!.where((rx) => rx.hashCode == ownerCode).isNotEmpty;
      if (!contains) {
        return;
      }
    }

    if (type == StateXType.isLoading) {
      var status = (state as StateXLoading).inLoading;
      _inLoading = status;
      if (widget.onLoading != null && status) {
        setState(() {
          _currentResult = widget.onLoading!();
        });
      } else {
        setState(() {
          _currentResult = widget.onState();
        });
      }
    } else if (type == StateXType.isError) {
      if (widget.onError != null) {
        setState(() {
          _currentResult = widget.onError!(state as Exception);
        });
      } else {
        setState(() {
          _currentResult = widget.onState();
        });
      }
    } else {
      if (!_inLoading) {
        setState(() {
          _currentResult = widget.onState();
        });
      }
    }
  }

  @override
  void initState() {
    _currentResult = widget.onState();
    widget.store.addObserver(_observer);
    super.initState();
  }

  @override
  void dispose() {
    widget.store.removeObserver(_observer);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return _currentResult;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:state_x/state_x.dart';

//Repositorio mockado para exemplificação
class Repository {
  Future<bool> authUser(String email, String passwold) async {
    await Future.delayed(Duration(seconds: 3));
    return true;
  }
}

//Dessa forma eu é criado um novo view model onde será centralizado do gerenciamento de estado da home
class HomeViewModel extends StateXStore {
  //Dessa maneira é criado um novo atribudo a ter o estado gerenciado pelo StateX
  late final Stated<String> xEmail = StateX.of(this)("");
  late final xPassword = StateX.of(this)("");
  late final xUserStatus = StateX.of(this)("Faça o login");

  final repo = Repository();

  set email(String value) => xEmail.value = value;
  String get email => xEmail.value;

  set password(String value) => xPassword.value = value;
  String get password => xPassword.value;

  set userStatus(String value) => xUserStatus.value = value;
  String get userStatus => xUserStatus.value;

  bool get isValid => email.isNotEmpty && password.isNotEmpty;

  void auth() async {
    //Aqui é informado que a view estará em loading
    setLoading(true);
    if (await repo.authUser(email, password)) {
      xUserStatus.value = "Usuário logado";
      //Aqui é informado que a view que o loading pode ser revogado
      setLoading(false);
    } else {
      //Em caso de erro assim deve ser setado a exceção
      setError(Exception("Email or password is invalid"));
    }
  }
}

void main() {
  //Aqui observer para todos os estados da aplicação util para gerenciamento de erros ou logs dos estados da aplicação
  StateXGlobalObserver.listen((type, state) => print(type));
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'StateX',
      theme: ThemeData(primarySwatch: Colors.blue, accentColor: Colors.white),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  var viewModel = HomeViewModel();

  @override
  void initState() {
    //Aqui é adicionado um novo observer util para gerenciamento de erros, contudo qualquer estado que seja mudado dentro da store passará por aqui
    viewModel
        .addObserver((type, state, key) => print(key == viewModel.xEmail.key));
    super.initState();
  }

  @override
  dispose() {
    viewModel.dispose();
    super.dispose();
  }

  Widget _buildButton(bool inLoading) {
    Widget buttonChild = inLoading
        ? SizedBox(
            width: 20,
            height: 20,
            child: CircularProgressIndicator(
              color: Colors.white,
            ))
        : Text("Enter");
    return Container(
      width: double.infinity,
      height: 48,
      child: ElevatedButton(
        onPressed: viewModel.isValid ? viewModel.auth : null,
        child: buttonChild,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("StateX"),
        ),
        body: Padding(
          padding: const EdgeInsets.all(21),
          child: Center(
            child: SingleChildScrollView(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  //Exemplo usando  scoped builder
                  StateXScopedBuilder(
                    //Informo qual a store que gerenciará as mudanças nesse escopo
                    store: viewModel,
                    //Parametro opcional. Especifica qual atributos irá escutar as mudanças de estado, caso não passe este atributo o scoped rebuildará a cada mudança da store
                    states: [viewModel.xUserStatus],
                    onState: () => Text(
                      viewModel.userStatus,
                    ),
                  ),
                  TextField(
                    onChanged: (value) {
                      viewModel.email = value;
                    },
                  ),
                  SizedBox(
                    height: 20,
                  ),
                  TextField(
                    onChanged: (value) {
                      viewModel.password = value;
                    },
                  ),
                  SizedBox(
                    height: 20,
                  ),

                  //Exemplo usando  scoped builder
                  StateXScopedBuilder(
                    //Informo qual a store que gerenciará as mudanças nesse escopo
                    store: viewModel,
                    //Parametro opcional. Especifica qual atributos irá escutar as mudanças de estado, caso não passe este atributo o scoped rebuildará a cada mudança da store
                    states: [viewModel.xEmail, viewModel.xPassword],
                    onLoading: () => _buildButton(true),
                    onState: () => _buildButton(false),
                  ),
                  SizedBox(
                    height: 20,
                  ),
                  //Exemplo usando state builder
                  StateXBuilder(
                    //Informo qual a store que gerenciará as mudanças nesse builder
                    store: viewModel,
                    //Parametro opcional. Especifica qual atributos irá escutar as mudanças de estado, caso não passe este atributo o scoped rebuildará a cada mudança da store
                    states: [viewModel.xEmail, viewModel.xPassword],
                    //Parametro informo que será rebuildado quando a store estiver em loading
                    includeLoading: true,
                    builder: (context, snapshot) {
                      return _buildButton(
                          snapshot.type == StateXType.isLoading);
                    },
                  )
                ],
              ),
            ),
          ),
        ));
  }
}
